[5.5] Sync Str with Stringy 3.0.1

fernandobandeira opened the pull request:
This PR intends to sync our core Str class with the behavior provided by Stringy library.

There's a few things that we should consider:

``` @ ``` was removed from the library so it no longers changes to ``` at ```, the library manually changes ``` @ ``` to ``` - ``` when generating the slug, I kept the previous behavior and moved it to the slug function, this way we can override the array in the future without caring about this and we also don't change the current behavior.

The library now uses locale's to replace characters according to the specific language. This also fixes some issues we had in the past of incorrect behavior, for example:

``` ü ``` previously was being replaced by ``` ue ``` now it is replaced by ``` u ``` and is replaced by ``` ue ``` just when using the ``` de ``` locale.

fernandobandeira committed:
Synced Stringy to 3.0.1

fernandobandeira committed:
Add language support for ascii

fernandobandeira committed:
Fix Style

fernandobandeira committed:
Beautify lang check

fernandobandeira committed:
Fix styleci

fernandobandeira committed:
Strict check

iwex commented on src/Illuminate/Support/Str.php:
```diff
@@ -33,10 +33,17 @@ class Str
     {
+        $languageSpecific = static::languageSpecificCharsArray($language);
+
+        if ($languageSpecific !== null) {
```
!is_null() seems more readable

fernandobandeira replied:
Changed 👍

iwex commented on src/Illuminate/Support/Str.php:
```diff
@@ -522,59 +532,85 @@ protected static function charsArray()
+     */
+    protected static function languageSpecificCharsArray($language)
+    {
+        static $languageSpecific;
```
original lib has more actions https://github.com/danielstjules/Stringy/blob/master/src/Stringy.php#L1857

fernandobandeira committed:
Readability changes

iwex commented:
@taylorotwell it's WIP, and lacks of additional actions like https://github.com/danielstjules/Stringy/blob/master/src/Stringy.php#L1863

fernandobandeira commented:
@iwex I've done that https://github.com/laravel/framework/pull/18974/files#diff-d19e1b0b8c742f268aabdbf1595bcfbdR614 just simplified it a bit.

fernandobandeira commented:
@iwex I've done that I just simplified the function https://github.com/laravel/framework/pull/18974/files#diff-d19e1b0b8c742f268aabdbf1595bcfbdR614

The behavior is the same ``` $charsArray ``` is a local static variable with all languages that are loaded, it contains the languages that were used before and add's new entries to the array whenever you use a new one, i removed this since there's just one language it would be overkill right now...

You might have confused it with the other ``` $charsArray ``` variable that contains all characters...

iwex commented:
@fernandobandeira thanks for explanation :) Stringy has really messy method

tillkruss commented on src/Illuminate/Support/Str.php:
```diff
@@ -522,59 +532,85 @@ protected static function charsArray()
+     *
+     * Note: Adapted from Stringy\Stringy.
+     *
+     * @see https://github.com/danielstjules/Stringy/blob/3.0.1/LICENSE.txt
```
Is this link to the licence intentional, or should this link to the source file?

fernandobandeira replied:
I copied it from the other method that graham created, I guess it's intentional to show that we can freely use this code on our project.
